src/sqlfluff/core/parser/lexer.py

```diff
@@ -322,9 +330,111 @@ def lex(self, raw: str) -> Tuple[Tuple[RawSegment, ...], List[SQLLexError]]:
                     # If we STILL can't match, then just panic out.
                     raise violations[-1]
 
-                raw = resort_res.new_string
+                str_buff = resort_res.new_string
                 start_pos = resort_res.new_pos
                 segment_buff += resort_res.segments
             else:
                 break
-        return segment_buff, violations
+
+        # Enrich the segments if we can using the templated file
+        return self.enrich_segments(segment_buff, raw), violations
+
+    @staticmethod
+    def enrich_segments(segment_buff, templated_file):
+        """Enrich the segments using the templated file.
+
+        We use the mapping in the template to provide positions
+        in the source file.
+        """
+        # Check this is actually a templated file.
+        if not isinstance(templated_file, TemplatedFile):
```

Comment:
should that check go outside of the function?

The arg and the docs specify that this function takes a `TemplatedFile` but it has the "hidden" behavior of returning the input unchanged if it's not

Reply:
Good idea 👍